src/viper/README.md

```diff
@@ -0,0 +1,141 @@
+# Formal Motoko
+
+**Disclaimer** — this is an _early prototype_; in particular: 
+1. This project comes with no guarantees whatsoever. 
+2. The currently supported subset of Motoko is insufficient for most real-world applications
+3. DFINITY currently has no plans to continue the development of Formal Motoko
+
+Formal Motoko is a prototype code-level verifier for Motoko. The project started at the DFINITY Foundation as a way to demonstrate that Motoko (and the Internet Computer in general) are well-suited for developing formally verified Web3 software.
+
+---
+**Jump to:**
+
+**[Introduction](#Introduction) —**
+  [Overview](#Overview)
+  | [Formal specification](#FormalSpecs)
+  | [Static vs. dynamic assertions](#statVsDyn)
+  | [Testing vs. formal verification](#testsVsVerif)
+  | [Precondition of public functions](#publicFuncPreconds)
+  | [Examples](#Examples)
+
+**[Contributing](#Contrib) —**
+    [Building](#Building)
+  | [Running](#Running)
+  | [Testing](#Testing)
+  | [File structure](#Struct)
+  | [Further information](#Further)
+
+---
+
+## Introduction
+<a name="Intro"></a>
+
+**Overview**
+<a name="Overview"></a>
+
+The verifier is implemented as a `moc` compiler feature. When `moc --viper $FILE` is invoked, the Motoko source code from `$FILE` is first translated into the Viper intermediate verification language (currently emitted into stdout) and then compiled. The generated Viper program can be submitted to the Viper tool. If the generated Viper program verifies, this implies that the Motoko source code respects the _formal code specification_, e.g., canister invariants, and that Viper established the corresponding formal proof of correctness.
+
+**Formal code specification**
+<a name="FormalSpecs"></a>
+
+Formal code specifications are written as part of the Motoko source code. These specifications are _static_, i.e., they do not affect the canister's runtime behavior. If a Motoko canister is compiled with `moc --viper`, it may contain static code specifications defined via the following keywords (`exp : Type` below denotes an expression of type `Type`):
+
+* `assert:invariant (exp : Bool);` — actor-level invariant; must be ensured right after canister initialization and after every _atomic block_ execution. May appear only at the level of actor members. 
+* `assert:1:async (exp : Bool);` — an `await async { ... }` block may include this as the first statement; the intention is two-fold:
+  * specify that the property `exp` is _intended_ to hold when this block execution begins. 
+  * require that the tool actually _checks_ whether this assumption holds (given this actor's entire source code)
+* `assert:system (exp : Bool);` — a _static assertion_ that asks the verifier to prove that the property `exp` holds. Useful while designing code-level canister specifications.
+
```

Comment:
Maybe note that above syntax is provisional (placeholder for potential bespoke Motoko syntax).

Reply:
```suggestion
Note: the above syntax is provisional. It has been used so far to avoid introducing breaking changes to the Motoko grammar. In the future, Formal Motoko may switch to bespoke syntax for code specifications. 

```